src/core/thread/multicast_listener_table.hpp

```diff
@@ -0,0 +1,215 @@
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file includes definitions for Multicast Listener Table on a Thread device.
+ */
+
+#ifndef MULTICAST_LISTENER_TABLE_HPP_
+#define MULTICAST_LISTENER_TABLE_HPP_
+
+#include "openthread-core-config.h"
+#include <common/locator.hpp>
+
+#if (OPENTHREAD_FTD || OPENTHREAD_MTD) && OPENTHREAD_CONFIG_MLR_ENABLE
+
+#include "common/code_utils.hpp"
+#include "common/debug.hpp"
+#include "net/ip6_address.hpp"
+
+namespace ot {
+
+/**
+ * @addtogroup core-mlr
+ *
+ * @{
+ *
+ */
+
+/**
+ * This class represents a Multicast Listener entry.
+ *
+ */
+class MulticastListener
+{
+    friend class MulticastListenerTable;
+
+public:
+    const Ip6::Address &GetAddress() const { return mAddress; }
+
+    bool IsDirty(void) { return mDirty; };
+    bool IsSubscribed(void) { return mSubscribed; }
+
+private:
+    MulticastListener()
+        : mSubscribed(false)
+    {
+        mAddress.Clear();
+    }
+
+    void Clear()
+    {
+        mAddress.Clear();
+        this->mSubscribed = false;
+    }
+
+    void SetSubscribed(bool aSubscribed) { mSubscribed = aSubscribed; }
+    void SetDirty(bool aDirty) { mDirty = aDirty; }
+
+    void SetAddress(const Ip6::Address &aAddress)
+    {
+        OT_ASSERT(mAddress.IsUnspecified());
+        OT_ASSERT(aAddress.IsMulticastLargerThanRealmLocal());
+
+        mAddress = aAddress;
+    }
+
+    bool IsDummy() { return !mSubscribed; }
+
+    Ip6::Address mAddress;
+    bool         mSubscribed : 1;
+    bool         mDirty : 1;
```

Comment:
add comment to illustrate the functionality of the fields to ease the review process?